### PR TITLE
Scheduler as a static pod

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
This is first step in making scheduler 
- As a static pod similar to apiserver and etcd.
- Breaking it from controllers.

The master config is still going to be similar, sample scheduler config is going to look like this:

https://gist.github.com/ravisantoshgudimetla/abb9c2d869aa54141c1b1a0a8e782859#file-scheduler-yaml

Note to myself: I need to create a ansible template for kube-scheduler.

/cc @deads2k @sjenning @aveshagarwal

### CODE REVIEW

File: pkg/cmd/server/start/start_scheduler.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	Start the master scheduler
+
+	This command starts the scheduler for the master.  Running
+
+	    %[1]s start master %[2]s
+
+	will start the the scheduler. The scheduler will run in the foreground until you terminate the process.`)
+
+// NewCommandStartMasterScheduler starts only the scheduler
+func NewCommandStartMasterScheduler(name, basename string, out, errout io.Writer) (*cobra.Command, *MasterOptions) {
```
Comment (deads2k) [inline pkg/cmd/server/start/start_scheduler.go]: We are trying to remove `openshift start *`.  I'd rather not see this added here.  What about `hyperkube kube-scheduler` with the right args?

**Comment by ravisantoshgudimetla** [comment]
@deads2k This is take2 on making scheduler as static pod - this time using `hyperkube scheduler` fetching arguments from masterconfig and use it for hyperkube image. 

The pattern is same as what are doing for node currently. I added a shim binary which converts master config to scheduler args(this could be expanded to args for other components). 

After the binary has been run we would get output like this:

`
--kubeconfig=/tmp/etc/origin/master/openshift-master.kubeconfig --policy-config-file=/etc/origin/master/scheduler.json --kube-api-content-type=application/vnd.kubernetes.protobuf --kube-api-qps=300 --kube-api-burst=600 --leader-elect=true --leader-elect-resource-lock=configmaps --port=0`

which could be passed onto the hyperkube scheduler as args. 

This the the commit to look at: https://github.com/openshift/origin/pull/20397/commits/611a559ebf0815bd21ecc6d0c3f3d11fbc99ace8

**Comment by deads2k** [comment]
This roughly the shape of what I was thinking.  Can you wire it up in cluster-up to prove it out.  Track down the spot from here: https://github.com/openshift/origin/blob/master/pkg/oc/clusterup/run_self_hosted.go#L66-L69

**Comment by ravisantoshgudimetla** [comment]
>This roughly the shape of what I was thinking. Can you wire it up in cluster-up to prove it out. Track down the spot from here

So, I am planning to test out in a master+3 node 3.10 cluster by adding my controlplane image(which doesn't start scheduler) and hyperkube image with scheduler as static pod. I could test it with `oc cluster up` as well.

**Comment by ravisantoshgudimetla** [comment]
@deads2k - I tested it with my 3 node cluster and scheduler from my origin-hyperkube image that I built locally is able to schedule pods with the above changes.

**Comment by mfojtik** [comment]
@ravisantoshgudimetla is still think that proving it in cluster up would be useful (at least it will be tested with every PR)

File: cmd/openshift-master-config/openshift-master-config.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/spf13/cobra"
+
+	"k8s.io/apiserver/pkg/util/logs"
+
+	"github.com/openshift/origin/pkg/cmd/flagtypes"
+	configapi "github.com/openshift/origin/pkg/cmd/server/apis/config"
+	configapilatest "github.com/openshift/origin/pkg/cmd/server/apis/config/latest"
+	configapiv1 "github.com/openshift/origin/pkg/cmd/server/apis/config/v1"
+	"github.com/openshift/origin/pkg/cmd/server/start"
+	"regexp"
```
Comment (mfojtik) [inline cmd/openshift-master-config/openshift-master-config.go]: nit: move these up

File: cmd/openshift-master-config/openshift-master-config.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	"k8s.io/apiserver/pkg/util/logs"
+
+	"github.com/openshift/origin/pkg/cmd/flagtypes"
+	configapi "github.com/openshift/origin/pkg/cmd/server/apis/config"
+	configapilatest "github.com/openshift/origin/pkg/cmd/server/apis/config/latest"
+	configapiv1 "github.com/openshift/origin/pkg/cmd/server/apis/config/v1"
```
Comment (mfojtik) [inline cmd/openshift-master-config/openshift-master-config.go]: @deads2k we haven't moved this to openshift/api yet?

**Comment by ravisantoshgudimetla** [comment]
> @ravisantoshgudimetla is still think that proving it in cluster up would be useful (at least it will be tested with every PR)
@mfojtik - Sure, I can test it. I think, I need to change the https://github.com/openshift/origin/blob/master/install/kube-scheduler/kube-scheduler.yaml#L16 to match my changes at https://gist.github.com/ravisantoshgudimetla/52f996d01d71dd5ea0a005073027a638

File: cmd/openshift-master-config/openshift-master-config.go (hunk 2 above)
Comment (ravisantoshgudimetla) [reply cmd/openshift-master-config/openshift-master-config.go]: Sure.

**Comment by ravisantoshgudimetla** [comment]
Closing this, as we are already scheduler as a separate pod(deployment) in 4.0.

@deads2k IIUC, for upgrade from 3.11 to 4.0, we need to translate master-config.yaml into a valid operator state, I think you guys are not working on it but you have planned for it. Please let me know, if 611a559 is something you could use for the translation code, if yes, I will open a PR once we are clear on where this should reside.

**Pull request closed without merging**